### DISCUSSION TITLE
validateOnly validates all matching fields in a collection

### DISCUSSION THREAD

**Discussion by acurrieclark** [livewire/livewire#3103] [7. Bugs (archived)]
### Description
When using validate only, I would expect that only the specific field would be validated against the rules. For example, if there is a collection of 3 `Reference` models, each with a `name` and `email` field, and the validation rules are:

```php
protected $rules = [
    'references.*.name' => ['required', 'distinct'],
    'references.*.email' => ['required', 'email', 'distinct'],
];
```

then when `$this->validateOnly('references.0.name')` is called, we should only validate the first reference's name. Instead, all 3 reference names will be validated, and an error added for each.

if `$this->validateOnly('references.*.name')` is called, the same behaviour applies, but this would be the expected result, to my mind.

### Exact steps to reproduce

I have made a [repo](https://github.com/acurrieclark/livewire-validation-tests) to reproduce the behaviour. Instructions in the readme:
`git clone git@github.com:acurrieclark/livewire-validation-tests.git` to take a look

**Context**
- Livewire version: 2.4
- Laravel version: 8.12
- Alpine version: n/a
- Browser: Brave/Safari/Firefox

Happy to put together a pull request for this if wanted.

**Comment by acurrieclark** [comment]
I have done a little more investigation into this in the last few hours and drew up a few tests to fit what I am suggesting.

In doing so, I found that there are actually 2 tests which already mark the current behaviour as correct. Specifically:

[old_deeply_nested_specific_validation_only_is_cleared_if_new_validation_passes](https://github.com/livewire/livewire/blob/b647ac63378ea02bdac975f70a4e57aea84bda0b/tests/Unit/ValidationTest.php#L213)
validates "items.0.baz" and then checks for an error on "items.1.baz"

[old_deeply_nested_specific_validation_only_is_cleared_if_new_validation_fails](https://github.com/livewire/livewire/blob/b647ac63378ea02bdac975f70a4e57aea84bda0b/tests/Unit/ValidationTest.php#L226)
validates "items.0.baz" and then checks that an error on "items.1.baz" has cleared.

What I am suggesting would mean that both of these tests would fail. So the above suggestion would be a breaking change.

**Comment by zepfietje** [comment]
@acurrieclark, I don't think the tests you linked to should fail, since they call the following method.
https://github.com/livewire/livewire/blob/b647ac63378ea02bdac975f70a4e57aea84bda0b/tests/Unit/ValidationTest.php#L464-L472

This tests exactly what you described here:
> if $this->validateOnly('references.*.name') is called, the same behaviour applies, but this would be the expected result, to my mind.

Without diving into the source I'm not 100% sure what causes this bug, but it's possibly an unintended side effect of https://github.com/livewire/livewire/pull/1615.

If you're able to write a couple of failing tests for the behavior you think should be different in this case, that'd be a nice first step towards fixing the issue (while not breaking the intended behavior of https://github.com/livewire/livewire/pull/1615).

**Comment by acurrieclark** [comment]
@zepfietje thanks for the review. I have put up a pull request with 2 added tests; one which tests the change I am proposing and fails and a second which tests the expected wildcard check and passes.

I don't think that what I am proposing would effect what is suggested in #1615, as long as the wildcard is submitted as the parameter for `validateOnly`.

One thing I did spot though, is that #1615 seems to validate (for example) `foo.*` as well as `foo.*.bar` if `foo.1.bar` is validated. It seems to me that `validateOnly` should only validate the specific field and leave all others untouched?

**Comment by zepfietje** [comment]
> I don't think that what I am proposing would effect what is suggested in #1615, as long as the wildcard is submitted as the parameter for `validateOnly`.

Yeah, but I can't remember if it's on purpose that all `items.*.foo` get validated when calling `validateOnly('items.0.foo')`.

> One thing I did spot though, is that #1615 seems to validate (for example) `foo.*` as well as `foo.*.bar` if `foo.1.bar` is validated. It seems to me that `validateOnly` should only validate the specific field and leave all others untouched?

So this is exactly what I'm unsure about. Also, I already noted `foo.*` getting validated too in a comment on that PR https://github.com/livewire/livewire/pull/1615#issuecomment-694149855.

**Comment by nuz** [comment]
Hi all,

I have the same problem. Rule is for persons.*.first_name. When validatesOnly() should validate persons.0.first_name, then persons.1.first_name and persons.2.first_name are also validated and errors appear. 
If I put in rules persons.0.first_name, persons.1.first_name and persons.2.first_name then validation works as it should. 

Problem is that persons are dynamic and it can be up to 6 of them.

Any solutions? @acurrieclark @zepfietje

**Comment by zepfietje** [comment]
@nuz, I don't have a workaround, but if @acurrieclark finds the time to finish his PR that'd fix it.

@acurrieclark, I think it should be safe to assume the behavior you described is what we want (not 100% sure though until Caleb has had a look at it).

**Comment by acurrieclark** [comment]
I have put some work into a PR, but I was waiting from some feedback from Caleb before I put anything up. There may be some thinking I am not aware of.

I do have a bit of a workaround which is based on the PR i am preparing, but it involves overriding the `validateOnce` method in your Livweire Component`.

```php
public function validateOnly($field, $rules = null, $messages = [], $attributes = [])
{
    $validErrors = collect($this->getErrorBag()->keys())->push($field);

    try {
        return parent::validateOnly($field, $rules, $messages, $attributes);
    } catch (ValidationException $e) {
        $newMessages = $e->validator->getMessageBag();
        $target = new ObjectPrybar($e->validator);

        $target->setProperty(
            'messages',
            new MessageBag(
                collect($newMessages->toArray())
                    ->filter(fn ($message, $key) => $validErrors->contains($key))
                    ->toArray()
            )
        );

        throw $e;
    }
}

```

**Comment by Tesla2k** [comment]
If solved that issue for me with nullable. But overall I think this behavior should be changed, it took me a while to find out what was going on.

```
protected $rules = [
    'data.*.value' => 'numeric|nullable',  
];
```
